File: variations.py
from __future__ import annotations

import pandas as pd


def compute_variations(df: pd.DataFrame) -> dict[str, float | None]:
    close = df["close"]
    n = len(close)

    def pct_change(periods: int) -> float | None:
        if n <= periods:
            return None
        return (close.iloc[-1] / close.iloc[-1 - periods] - 1) * 100

    return {"d1": pct_change(1), "d7": pct_change(7), "d30": pct_change(30)}

File: pwa/js/variations.js
// Port de variations.py
function computeVariations(candles) {
  const closes = candles.map((c) => c.close);
  const n = closes.length;
  function pctChange(periods) {
    if (n <= periods) return null;
    return (closes[n - 1] / closes[n - 1 - periods] - 1) * 100;
  }
  return { d1: pctChange(1), d7: pctChange(7), d30: pctChange(30) };
}
